internal/sbi/httpcallback/api_handle_dereg_notification.go

```diff
@@ -0,0 +1,138 @@
+
+	var deregData models.DeregistrationData
+	var doUecmDereg bool = true
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		logger.CallbackLog.Errorf("Get Request Body error: %+v", err)
+		problemDetails := models.ProblemDetails{
+			Title:  "System failure",
+			Status: http.StatusInternalServerError,
+			Detail: err.Error(),
+			Cause:  "SYSTEM_FAILURE",
+		}
+		c.JSON(http.StatusInternalServerError, problemDetails)
+		return
+	}
+
+	err = openapi.Deserialize(&deregData, requestBody, "application/json")
+	if err != nil {
+		problemDetails := "[Request Body] " + err.Error()
+		rsp := models.ProblemDetails{
+			Title:  "Malformed request syntax",
+			Status: http.StatusBadRequest,
+			Detail: problemDetails,
+		}
+		logger.CallbackLog.Errorln(problemDetails)
+		c.JSON(http.StatusBadRequest, rsp)
+		return
+	}
+
+	ueid := c.Param("ueid")
+	amfSelf := amf_context.GetSelf()
+	ue, ok := amfSelf.AmfUeFindByUeContextID(ueid)
+	if !ok {
+		logger.CallbackLog.Errorf("AmfUe Context[%s] not found", ueid)
+		problemDetails := &models.ProblemDetails{
+			Status: http.StatusNotFound,
+			Cause:  "CONTEXT_NOT_FOUND",
+		}
+		c.JSON(http.StatusNotFound, problemDetails)
+		return
+	}
+
+	switch deregData.DeregReason {
+	case models.DeregistrationReason_UE_INITIAL_REGISTRATION:
+		// TS 23.502 - 4.2.2.2.2 General Registration
+		// Invokes the Nsmf_PDUSession_ReleaseSMContext for the corresponding access type
+		ue.SmContextList.Range(func(key, value interface{}) bool {
+			smContext := value.(*amf_context.SmContext)
+
+			if smContext.AccessType() == deregData.AccessType {
+				var problemDetails *models.ProblemDetails
+				problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
+				if problemDetails != nil {
+					ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetails)
+				} else if err != nil {
+					ue.GmmLog.Errorf("Release SmContext Error[%v]", err.Error())
+				}
+			}
+
+			return true
+		})
+	case models.DeregistrationReason_SUBSCRIPTION_WITHDRAWN:
+		// The notification is triggered by the new AMF executing UECM registration,
+		// UDM will use the new data to replace the old context.
+		// Therefore old AMF doesn't need to execute UECM de-registration to clean the old context stored in UDM.
+		doUecmDereg = false
+	}
+	// TS 23.502 - 4.2.2.2.2 General Registration - 14e
+	// TODO: (R16) If old AMF does not have UE context for another access type (i.e. non-3GPP access),
+	// the Old AMF unsubscribes with the UDM for subscription data using Nudm_SDM_unsubscribe
+	if ue.SdmSubscriptionId != "" {
+		var problemDetails *models.ProblemDetails
+		problemDetails, err = consumer.SDMUnsubscribe(ue)
+		if problemDetails != nil {
+			logger.GmmLog.Errorf("SDM Unubscribe Failed Problem[%+v]", problemDetails)
+		} else if err != nil {
+			logger.GmmLog.Errorf("SDM Unubscribe Error[%+v]", err)
+		}
+		ue.SdmSubscriptionId = ""
+	}
+
+	if doUecmDereg {
```

Comment:
Yes, it will

Reply:
So should UE_INITIAL_REGISTRATION also exclude UeCmDereg?

Reply:
From my understanding, if the reason is UE_INITIAL_REGISTRATION the consumer NF should be during the initial registration procedure. Therefore I think the consumer NF should execute the UECM Deregistration.
For example, let's say the new AMF has completed the UECM Registration but has not completed the initial registration procedure yet. If it receives the deregistration notification from UDM, I think the reason may be UE_INITIAL_REGISTRATION.
Another example is that the old AMF has completed the initial registration procedure and the new AMF is during the initial registration procedure. The UDM sends a deregistration notification with the reason SUBSCRIPTION_WITHDRAWN to the old AMF.
If I was wrong, please correct me.

Reply:
According to 23.502 4.2.2.2.2, I think only the old AMF receives UECM_DeregistrationNotification during the initial registration, and I think UE_INITIAL_REGISTRATION is the only dereg reason that fits it🤔. (Or did you see any other parts of specs mentioning this part?)
![image](https://github.com/free5gc/amf/assets/18053123/a164e50e-ca6b-4662-956a-bb8b26dcc35b)

For the reason SUBSCRIPTION_WITHDRAWN, I saw it mentioned in 4.2.2.3.3 Network-initiated Deregistration.
> 1. [Conditional] If the UDM wants to request the immediate deletion of a subscriber's RM contexts and PDU 
Sessions, the UDM shall send a Nudm_UECM_DeregistrationNotification (SUPI, Access Type, Removal 
Reason) message with Removal Reason set to Subscription Withdrawn to the registered AMF. The Access Type 
may indicate 3GPP Access, non-3GPP Access or both.  
Indeed, I'm not quite sure whether the cited parts include all cases for SUBSCRIPTION_WITHDRAWN, so feel free to open more discussions if you have any concerns🙏

Reply:
I saw the description of TS 23.502 - 5.2.3.2.2 Nudm_UECM_DeregistrationNotification service operation 
> Description: UDM notifies the NF consumer which has previously registered (using Nudm_UECM_Registration operation) has been deregistered in the UDM. As a result, the consumer is no longer registered in UDM as a serving NF for that UE.

Therefore, I think it is possible that the new AMF may receive the Nudm_UECM_DeregistrationNotification.

> According to 23.502 4.2.2.2.2, I think only the old AMF receives UECM_DeregistrationNotification during the initial registration, and I think UE_INITIAL_REGISTRATION is the only dereg reason that fits it🤔. (Or did you see any other parts of specs mentioning this part?)

I didn't see other specs mention how to set the dereg reason.
If UDM sends the dereg notification with UE_INITIAL_REGISTRATION, it implies the old AMF should initiate Nsmf_PDUSession_ReleaseSMContext toward all the associated SMF(s) of the UE, then SMF should release the PDU session on getting this notification.
However, I found that the Namf_Communication_UEContextTransfer is able to transfer some context related to the PDU session (e.g. PDU session ID), I am thinking if it means that after the UE register to the new AMF, it can directly use the PDU session that was established through the old AMF. If yes, in this case, it is not proper that UDM send the dereg notification with UE_INITIAL_REGISTRATION.
Therefore, sending dereg notification with SUBSCRIPTION_WITHDRAWN seems to be the proper way, it only lets the old AMF no longer serve the UE.

Reply:
1. I don't quite understand why the first citation results in the conclustion that the new AMF still has a chance to receive the  Nudm_UECM_DeregistrationNotification😂. "UDM notifies the NF consumer which has previously registered" -> I think only an _old_ AMF can have a previous registration to UDM for the UE that is initiating a new registration to a _new_ AMF. In other words, if an AMF has previously registered to the UDM using Nudm_UECM_Registration, it must have a UeCtx for that UE and can't be called _new_ AMF when the UE registers to it again.
2. I checked for the PDU Session transferring parts in 23.502 and yeah...the parts shown in the images are confusing regarding initial registration.
![image](https://github.com/free5gc/amf/assets/18053123/98149195-5004-4b50-bf10-389364268c15)
Though not mentioned by spec, I personally think step 5. is mainly for mobility/periodic registration. The reason is that a UE will do initial registration only when it believes that it is in DEREGISTERED state, and no SM contexts should exist for a UE in DEREGISTERED state. In my opinion, I think the flow should be like the followings if old AMF holds pdu sessions contexts when doing UeCtxTransfer at UE's initial registration to new AMF (this scenario may happen due to e.g. previous deregistration procedure didn't complete accidently):
   - step 4. : New AMF's UeCtxTransfer request should set transfer reason to "INIT_REG" (Ref: 29.518 6.1.6.3.14)
   - step 5. :  Old AMF's should not fill PDU Sessions in UeCtxTransferRsp when the transfer reason is "INIT_REG"
   - step 14d. : Old AMF should release those old PDU Sessions with SMF since dereg reason is UE_INITIAL_REGISTRATION
   - step 17. : New AMF won't do SM Context Update to SMF since no Pdu Sessions are received in step 5.

In conslusion, I still believe that the dereg reason for initial registration can only be UE_INITIAL_REGISTRATION and UECM_Deregistration should not be sent by old AMF to UDM in this case.

Reply:
Hi @YouShengLiu @ss920386 

> I don't quite understand why the first citation results in the conclustion that the new AMF still has a chance to receive the Nudm_UECM_DeregistrationNotification😂. "UDM notifies the NF consumer which has previously registered" -> I think only an old AMF can have a previous registration to UDM for the UE that is initiating a new registration to a new AMF. In other words, if an AMF has previously registered to the UDM using Nudm_UECM_Registration, it must have a UeCtx for that UE and can't be called new AMF when the UE registers to it again.

I think the New AMF has the possibility to receive the Nudm_UECM_DeregistrationNotification.
For example:
After UE has registered to the new AMF after the registration procedure. if the operator decides to withdraw the subscription data of this UE (e.g. delete subscription data via webconsole), the UDM should send the Nudm_UECM_DeregistrationNotification with reason SUBSCRIPTION to the new AMF (current serving AMF to the UE).

Reply:
Anyway, I think the case of SUBSCRIPTION_WITHDRAWN won't happen in the free5GC. because we haven't implemented the feature of data changed notification in our UDR yet.

Reply:
@ianchen0119 Yes, it is possible for the new AMF to receive Nudm_UECM_DeregistrationNotification with SUBSCRIPTION_WITHDRAWN. What we tried to clarify is whether only the old AMF should receive  Nudm_UECM_DeregistrationNotification during **initial registration**, and whether SUBSCRIPTION_WITHDRAWN is possible to be the dereg reason in this case. Do you have any idea?😂

Reply:
My original thought was that the new AMF may have the possibility to receive the Nudm_UECM_DeregistrationNotification between step 14a. and step 22.
Now, I don't think so.
If the dereg reason is UE_INITIAL_REGISTRATION, the Nudm_UECM_DeregistrationNotification must be triggered when a new AMF handles the initial registration request for the UE. Only the old AMF will receive the notification.

As for the SUBSCRIPTION_WITHDRAWN, I think its use case should be @ianchen0119 mentioned. It is more reasonable.

Let me summarize what should be modified now:

- If the old AMF receives the Nudm_UECM_DeregistrationNotification with the reason UE_INITIAL_REGISTRATION, it should not execute UECM_Deregistration. Because the new AMF will take over it.
- If the old AMF receives the Nudm_UECM_DeregistrationNotification with the reason SUBSCRIPTION_WITHDRAWN, it should execute UECM_Deregistration. Because the subscriber has been removed from the core network, there would be no AMF can serve the UE.
- The dereg reason set from UDM needs to be modified, it should be set to UE_INITIAL_REGISTRATION rather than SUBSCRIPTION_WITHDRAWN.

Reply:
Yeah, I agree with that. Thank you and @ianchen0119  for all the discussions. In addition, could you please help to check if these two steps implementation is as expected when it is initial registration?
step 4. : New AMF's UeCtxTransfer request should set transfer reason to "INIT_REG" (Ref: 29.518 6.1.6.3.14)
step 5. : Old AMF's should not fill PDU Sessions in UeCtxTransferRsp when the transfer reason is "INIT_REG"

Reply:
Sorry, I have one more concern for
> If the old AMF receives the Nudm_UECM_DeregistrationNotification with the reason SUBSCRIPTION_WITHDRAWN, it should execute UECM_Deregistration. Because the subscriber has been removed from the core network, there would be no AMF can serve the UE.

after seeing these parts in TS 23.502:
![image](https://github.com/free5gc/amf/assets/18053123/b9d67fd1-2f72-4514-9966-970e3e84f99e)

Should Nudm_UECM_DeregistrationNotification and Nudm_UECM_Deregistration be mutually exclusive? Nudm_UECM_DeregistrationNotification tells AMF (no matter new or old) that it's no longer registered in UDM, so does that mean we just don't need to do Nudm_UECM_Deregistration, which requests for deregistering in UDM?

Reply:
Hi @YouShengLiu @ss920386 
I think @ss920386's concern is right, those service operations should be mutually exclusive.

Reply:
If that is the case, another concern is that if UDM does deregister the AMF when it sends UECM_DeregistrationNotification in our UDM implementation. (It is possible that our UDM simply sents UECM_DeregistrationNotification to AMF without handling the related contexts.)

Reply:
For the current implementation in UDM, it will use the Amf3GppAccessRegistration from the new AMF to replace the Amf3GppAccessRegistration from the old AMF.
Then, the UDM will send UECM_DeregistrationNotification to AMF.
Amf3GppAccessRegistration is the context related to the Nudm_UECM for AMF registration for 3GPP access.

Reply:
@YouShengLiu How about SUBSCRIPTION_WITHDRAW?

Reply:
Although the current UDM sends the dereg notification with the reason SUBSCRIPTION_WITHDRAW, based on our discussion, I think it should be modified to UE_INITIAL_REGISTRATION.
Therefore, the current UDM would not have the possibility to send the dereg notification with the reason SUBSCRIPTION_WITHDRAW.
As for further development, we may implement another API to handle this situation (e.g. OAM) or it will be triggered from UDR when the subscription data is removed.